assets/src/block-editor/QueryBlockExtension.js

```diff
@@ -0,0 +1,140 @@
+import {ToolbarGroup, ToolbarButton, RadioControl, PanelBody} from '@wordpress/components';
+import {InspectorControls, BlockControls} from '@wordpress/block-editor';
+import {useCallback, useEffect, useMemo} from '@wordpress/element';
+
+const {addFilter} = wp.hooks;
+const {__} = wp.i18n;
+
+const targetBlocks = [
+  'core/query',
+];
+
+const layoutOptions = [
+  {label: 'List', value: 'list'},
+  {label: 'Grid', value: 'flex'},
+  {label: 'Carousel', value: 'carousel'},
+];
+
+export const setupQueryBlockExtension = function() {
+  const {createHigherOrderComponent} = wp.compose;
+
+  addFilter(
+    'editor.BlockEdit',
+    'planet4-blocks/overrides/query-controls',
+    createHigherOrderComponent(
+      BlockEdit => props => {
+        if (!targetBlocks.includes(props.name)) {
+          return (
+            <BlockEdit {...props} />
+          );
+        }
+
+        const {attributes, setAttributes} = props;
+
+        const onChangeHandler = useCallback(view => {
+          let layoutType = 'list';
+          let layoutColumns = 3;
+          let queryPerPage = 3;
+
+          switch (view) {
+          case 'flex':
+            layoutType = 'flex';
+            break;
+          case 'carousel':
+            layoutType = 'carousel';
+            layoutColumns = 6;
+            queryPerPage = 6;
+            break;
+          }
+
+          setAttributes({
+            displayLayout: {
+              type: layoutType,
+              columns: layoutColumns,
+            },
+            query: {
+              ...attributes.query,
+              perPage: queryPerPage,
+            },
+          });
+        }, []);
+
+        useEffect(() => {
+          const pattern = 'is-custom-layout-';
+          const {className, displayLayout} = attributes;
+
+          if (typeof className !== 'undefined') {
+            setAttributes({
+              className: (className.includes(pattern)) ?
+                className.replace(/\is-custom-layout-.*/, `${pattern}${displayLayout.type}`) :
+                `${className} ${pattern}${displayLayout.type}`,
+            });
+          }
+        }, [
+          attributes,
+          setAttributes,
+        ]);
+
+        useEffect(() => {
+          if (props.isSelected) {
+            const timeout = setTimeout(() => {
+              // Move the carousel view button into the group of views
```

Comment:
Need to implement this tricky code to move the carousel layout option to the proper group

Before 
![Screenshot 2023-12-07 at 12 06 37](https://github.com/greenpeace/planet4-master-theme/assets/77975803/837b37a3-4bd6-4ecb-9d34-cda7a61692e8)

After
![Screenshot 2023-12-07 at 12 24 01](https://github.com/greenpeace/planet4-master-theme/assets/77975803/1c19e092-b0c9-4ab0-bd3d-604ec823b593)

Reply:
Maybe it wasn't such a big deal if they were separated? I just think that this is a lot of complex code for such a small thing 🤔 Also in the final result there is now an empty section which looks weird as well, I would prefer to keep the version with the carousel separated!

Reply:
Sorry, I've attached a wrong screenshot. The empty section is not there anymore 👇 

<img width="669" alt="Screenshot 2023-12-12 at 06 18 57" src="https://github.com/greenpeace/planet4-master-theme/assets/77975803/d1c363ee-e00c-4440-886f-ef2d56eda898">

You can also test it on any of the shared test pages. Still disagree?

Reply:
Ah, nice 😄 I still think it's a bit complex & unnecessary, and it will likely cause problems with the WP 6.4 upgrade (I think that's what crashes on my local)
But I also understand that it took you some time and it does look better, so again it's just my opinion!

Reply:
I see your point about the complexity and I completely agree with that. In fact, I commented the same thing [here](https://github.com/greenpeace/planet4-master-theme/pull/2163#discussion_r1419491121). 
However, we will have inconsistencies between the toolbar and the right nav.. don't you think that won't be an "issue" to editors?
BTW, I can take a look what is wrong with `6.4`.

Reply:
I see your point, but I don't think it's a huge deal 🤔 For 6.4 I believe Flo is already looking into it so you can see with him, in fact he opened a [PR](https://github.com/greenpeace/planet4-master-theme/pull/2169) yesterday

Reply:
I'm a bit confused by this extension 🤔  This seems to add a Layout sidebar (and the Carousel style) on Query Loop block and all of its variations. But we only need the Carousel style in the Actions List block.

Reply:
Yes! In this case we should apply only to Actions List, but it's also applied to the rest of `core/query`.
I'm going to update the code then.